Improve proxy security (#100)

toplenboren opened the pull request:
Fixes #100 

I added some small improvements to `proxy.ts` service

I've implemented:
- Do not allow POST / PUT and pretty much everything besides GET
- Do not allow requests to reserved ips
- Do not allow anything besides http or https
- Remove shady headers
- Added a handy README note
- Test that init real target http server and proxy and test end to end user 

I asked a few questions in #100. Let's use issue to discuss planned changes

## Motivation

We do not want our proxy to be abused and banned, thus all the security stuff. Check #100 for details

## Deps

I needed a package to check whether IP is reserved or not. Thought that we do not want to bloat our code with this stuff. I found three packages: isMartianPacket, bogon and isReservedIp

I found [is-reserved-ip](https://www.npmjs.com/package/reserved-ip) code to be strange and [bogon](https://www.npmjs.com/package/bogon) to have one more dependency 👍 

I decided to pick [isMartianPacket](https://www.npmjs.com/package/martian-cidr).Package seems to be [maintained](https://github.com/arty-name/martian-cidr/issues/50), and I found the lack of types (Can implement this myself 🥨) and lack of adoption as main drawback (thou need to say that many people just introduce code for this task in their projects).  

## Checklist (Gonna be checked once we are out of draft state)

- [x] Don’t rush. Check all changes in PR again.
- [x] Run `pnpm test`.
- [x] Think about changing documentation.
  - If you added a script to `scripts/`, add a comment with a description.
  - If you added a new folder, add its description to the project’s `README.md`.
  - If you added config, describe how we use this tool in the config’s comment.
  - If you added something to the project’s architecture, describe it in the project’s `README.md`.
  - Try to focus on “why?”, not “how?”.
- [x] If you added a new dependency, check our [requirements](../README.md#dependencies).
- [x] Think about testing
  - If you added a feature, add unit tests.
  - If you added a new state to the UI, add visual tests.
  - If you fixed the bug, think about preventing bug regression in the future.
- If you changed web client:
  - [x] Think about moving code to `core/`. What code will also be useful on other platforms?
  - [x] Run `pnpm size` and check the difference in the JS bundle size. Is it relevant to the changes? Change the limit in `web/.size-limit.json` if necessary.
  - [x] Think about keyboard UX. Is it easy to use the new feature with only one hand on a keyboard? Is it easy to understand what keys to press?
  - [x] Think about HTML semantics.
  - [x] Think about accessibility. Check a11y recommendations. Think about how screen reader users will use the tool. Is it easy to use on a screen with bad contrast?
  - [x] Think about translations. Will
  - [x] Think about right-to-left languages. What parts of the screen should be mirrored for Arabic or Hebrew languages?
- If you changed the colors token in the web client:
  - [x] Think about app loading styles inlined in `index.html`.
- If you changed `core/`:
  - [x] Think about making types more precise. Can you better explain data relations by type?
  - [x] Think about conflict resolution. We don’t need some very smart changing merging; just 2 changes of the same item on different clients should not break the database. What if the user changes an item on one machine and removes it on another?
  - [x] Think about log and storage migration.
- If you changed English translations:
  - [x] Change translation ID if you change the meaning of the text.

ai commented on proxy/README.md:
````diff
@@ -3,3 +3,24 @@
+
+### Start proxy
+
+```shell
````
```suggestion
```sh
```

ai commented on proxy/README.md:
````diff
@@ -3,3 +3,24 @@
+### Make a request
+
+```shell
+curl localhost:5284/https://hplush.dev
````
Let’s not document the API 😏

toplenboren replied:
Alright :-)

toplenboren replied:
Done!

ai commented on proxy/package.json:
```diff
@@ -3,7 +3,7 @@
   "type": "module",
   "scripts": {
-    "start": "tsx ./index.ts"
+    "start": "tsx ./proxy.ts"
```
In all another project the main file is `index.ts` (like in `server/`). Let’s keep name system.

toplenboren replied:
Makes sense :-) let's keep it systematical

toplenboren replied:
Done!

ai commented on proxy/index.ts:
```diff
@@ -1,43 +1,19 @@
+const PRODUCTION_DOMAIN_SUFFIX = '.slowreader.app'
+
+const proxy = createProxyServer({
+  isProduction: IS_PRODUCTION,
```
Do we need constant if we use it right here?

toplenboren replied:
I just really like to put it on the top of the file in case we want to add additional logic on top of our proxy launch. 

By the way, what do you think about the way to configure proxy? 

We'd need to set process.env.mode to `production` on production.. How do you deploy this app? Is this kind of configuration ok for you?

ai replied:
I will do it when I will prepare deploy (it need to be done together with Google Cloud access for tests).

For instance, we will nee to read port from `env.PORT`, etc.

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,119 @@
+ *
+ * @param { isProduction, silentMode, hostnameWhitelist, productionDomainSuffix } config
+ */
+const createProxyServer = (
```
We do not use default export in the project (they sometimes have a problem in bundlers/tools, not a big dead but nice for code style—do we have ESLint rule for that?).

Let’s use `export function createProxyServer`

toplenboren replied:
Fixed, thanks!

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,119 @@
+    try {
+      // Only http or https protocols are allowed
+      if (!url.startsWith('http://') && !url.startsWith('https://')) {
+        throw new Error('Only http or https requests are allowed.')
```
It will be nice to have `400` (Bad request) response here. It seems for me that `Error` is always will return `500`,

toplenboren replied:
Agreed, added 400 response here

toplenboren replied:
Agreed, just fixed that

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,119 @@
+    } catch (e) {
+      if (e instanceof Error) {
+        if (!silentMode) {
+          process.stderr.write(styleText('red', e.stack ?? e.message) + '\n')
```
We should not write `400` errors and timeout errors to the log, but need to write all other errors.

toplenboren replied:
Agreed, just fixed that

toplenboren replied:
Agreed, just fixed that

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,119 @@
+          res.end('Internal Server Error')
+        }
+      } else if (typeof e === 'string') {
+        if (!silentMode) {
```
I think here it will be better to re-use `if (!sent)` branch (by moving it out of the `if (typeof e === 'string') {`

toplenboren replied:
Agreed. Code will smell better with this change! Just fixed that

ai commented on proxy/test/proxy.test.ts:
```diff
@@ -0,0 +1,163 @@
+  )
+})
+
+describe('proxy tests', async () => {
```
We don’t use single `describe` per file in other tests.

toplenboren replied:
Can we have a single describe here? The reason is that I want to utilise `before` and `after` so that I can init and destroy http server. 

If I ditch single describe, then I'd have to write a lot of:

```js
await initTestHttpServer(
    'proxy',
    createProxyServer({ hostnameWhitelist: ['localhost'], silentMode: true })
  )
  await initTestHttpServer('target', targetServer)

  let proxyServerUrl = ''
  let targetServerUrl = ''

  before(() => {
    let proxy = getTestHttpServer('proxy')
    if (proxy) {
      proxyServerUrl = proxy.baseUrl
    }

    let target = getTestHttpServer('target')
    if (target) {
      targetServerUrl = target.baseUrl
    }

    if (!target || !proxy) {
      throw new Error(
        "Couldn't set up target server or proxy server. Please check out 'proxy.test.js'"
      )
    }
  })
```

Maybe there is a way to support before outside the single describe?

ai replied:
You can use `before`/`after` without `description`.

Just use `beforeEach` or `beforeAll` from `node:test`.

ai commented on proxy/test/proxy.test.ts:
```diff
@@ -0,0 +1,163 @@
+
+    if (!target || !proxy) {
+      throw new Error(
+        "Couldn't set up target server or proxy server. Something is misconfigured. Please check out 'proxy.test.js'"
```
Can we make it shorter to fit `80` columns line?

toplenboren replied:
Just made that comment shorter! Thanks

ai commented on proxy/test/proxy.test.ts:
```diff
@@ -0,0 +1,163 @@
+  await test('proxy works', async () => {
+    let response = await fetch(`${proxyServerUrl}/${targetServerUrl}`)
+    let parsedResponse = await response.json()
+    // @ts-ignore
```
```suggestion
    // @ts-expect-error
```

BTW, why?

ai replied:
Maybe `equal(response?.status, 200)`

toplenboren replied:
Just fixed this, thanks!

ai commented on proxy/test/proxy.test.ts:
```diff
@@ -0,0 +1,163 @@
+    })
+
+    await test('can use only http or https protocols', async () => {
+      let forbiddenProtocols = [
```
A single test is enough :D

toplenboren replied:
I agree with you, but I think it is better to just add more protocols and don't think of it again :D

ai commented on proxy/test/utils.ts:
```diff
@@ -0,0 +1,76 @@
+  server: any,
+  opts?: { port?: number; protocol?: string }
+): Promise<void> {
+  //
```
```suggestion
```

toplenboren replied:
Just fixed that

ai commented on proxy/test/utils.ts:
```diff
@@ -0,0 +1,76 @@
+ */
+export async function initTestHttpServer(
+  name: string,
+  server: any,
```
Not cool :)

toplenboren replied:
Just fixed that!

ai commented on proxy/test/utils.ts:
```diff
@@ -0,0 +1,76 @@
+  [key: string]: TestServer
+} = {}
+
+/**
```
We do not need JSDoc here. It is better to add it everywhere or not use it.

toplenboren replied:
Just checked out other files and added a small commentary on the top of module instead of docstrings

ai commented on proxy/test/utils.ts:
```diff
@@ -0,0 +1,76 @@
+  server: http.Server
+}
+
+const __testServers: {
```
Why not just `testServers`?

toplenboren replied:
I usually like to indicate that this variable is private, but let's be simple here! Changed back to `testServers`

ai commented on proxy/test/utils.ts:
```diff
@@ -0,0 +1,76 @@
+
+    // port=0 is <any free port>
+    let testServerInstance = await server.listen(port)
+    // @ts-ignore
```
What is the error here?

toplenboren replied:
The error was in the way `.addressInfo()` data was shaped. I think I figured it out with type conversion

ai commented on proxy/README.md:
````diff
@@ -3,3 +3,18 @@
+# // Proxy server running on port 5284
+```
+
+## Additional Info:
````
Title should not contain `:`

toplenboren replied:
Just fixed `README.md`

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,119 @@
+          targetResponse.headers.get('content-type') ?? 'text/plain'
+      })
+      sent = true
+      res.write(await targetResponse.text())
```
Do you want to also improve a performance a little?

Why do we need to first create a full copy of the HTTP document in the proxy memory and only then send it to the user? Instead, on receiving first bites we can send them to the client. It improves performance and memory consumption.

https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Using_readable_streams

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,119 @@
+      res.writeHead(targetResponse.status, {
+        'Access-Control-Allow-Headers': '*',
+        'Access-Control-Allow-Methods': 'OPTIONS, POST, GET, PUT, DELETE',
+        'Access-Control-Allow-Origin': '*',
```
There is a cool idea of using not `*` but `req.headers['Origin']`

toplenboren replied:
Seems cool. Just added this feature

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,119 @@
+      let targetResponse = await fetch(url, {
+        headers: {
+          ...(req.headers as HeadersInit),
+          host: new URL(url).host
```
Should we write client’s IP to avoid using proxy to hide IP? There is `X-Forwarded-For` header for that.

Later we will add privacy mechanism for paid users.

toplenboren replied:
Agreed. Let me take some time to research usecases and write tests :D

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,119 @@
+      })
+      sent = true
+      res.write(await targetResponse.text())
+      res.end()
```
There is another nice abuse protection system. We can limit the size of response https://github.com/gridaco/cors.sh/blob/main/services/proxy.cors.sh/src/limit/payload-limit.ts

toplenboren commented on proxy/README.md:
```diff
@@ -3,3 +3,18 @@
 
 User could use it to bypass censorship or to try web client before they install upcoming extension (to bypass CORS limit of web app).
+
+## Usage
```
Fixed `README.md`

toplenboren commented:
Hey, @ai I've done some cool stuff! Can you please re-review my PR? 

## Done:

- Introduced timeout feature
- Refactored error handling: All Bad Requests will not be looged and will return 400 to user
- Fixed minor issues, such as README.md structure or typescript errors

## Questions:

- I have a question about single `describe` in tests. Please check it
- I have a question about configuration

## Left to do:

1. Debounce
2. Hide client IP
3. Limit size of response
4. Introduce partial return for optimizations

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,126 @@
+// @ts-ignore
```
Let’s define simple types in `global.d.ts`. It should be a few lines just for API which we use.

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,126 @@
+    silentMode?: boolean
+  } = {}
+): http.Server => {
+  // Main toggle for production features
```
The arguments are very simple to explain them in comments (in contrast with different hacks)

ai commented on proxy/README.md:
```diff
@@ -3,3 +3,23 @@
+## Abuse Protection
+
+- Proxy allows only GET requests
+- Proxy do not allow requests to reserved ip addresses like `127.0.0.1`
```
```suggestion
- Proxy do not allow requests to in-cloud IP addresses like `127.0.0.1`
```

ai commented on proxy/README.md:
```diff
@@ -3,3 +3,23 @@
+
+- Proxy allows only GET requests
+- Proxy do not allow requests to reserved ip addresses like `127.0.0.1`
+- Proxy allows only http or https
```
```suggestion
- Proxy allows only HTTP or HTTPS protocols
```

ai commented on proxy/README.md:
```diff
@@ -3,3 +3,23 @@
+
+## Test Strategy
+
+Proxy is tested using e2e testing. To write tests `initTestHttpServer` should be used
```
```suggestion
To test proxy we emulate the real HTTP servers (end-to-end testing).
```

Docs are for juniors.

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,126 @@
+  }
+}
+
+export const createProxyServer = (
```
Let’s use `function` if it is a function

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,126 @@
+
+      let requestUrl = new URL(url)
+      if (!hostnameWhitelist.includes(requestUrl.hostname)) {
+        // Do not allow requests to addresses that are reserved:
```
```suggestion
        // Do not allow requests to addresses inside our cloud:
```

We should explain “why we did it” not “what we did”

ai commented on proxy/proxy.ts:
```diff
@@ -0,0 +1,126 @@
+        }
+      }
+
+      // Remove all cookie headers and host header from request
```
```suggestion
      // Remove all cookie headers so they will not be set on proxy domain
      // Replace Host header to target
```

ai commented on proxy/test/utils.ts:
```diff
@@ -0,0 +1,68 @@
+//  Use this function if you need to init local http server for testing. Express is supported. Server will be closed automatically after tests finish
```
We don’t need docs here since the test is a good example
